README.md
Deebee (ß)
============
Deebee is a wrapper for Dropbox iOS SDK 1.3. It aims to emulate Dropbox's desktop client's behaviour and to add more filters to help developers achieve their apps' goals. As of this beta, Deebee was entirely developed by me, and I'm a designer and have known Objective-C for two months, so please, make pull requests if you believe there's something wrong. I'm eager to learn more, and I'd really appreciate your help.  
  
I built Deebee because I needed it for some apps I conceived and designed, so feel free to use it however you want: it's licensed with MIT license for the moment.  
It's important that you consider it as a beta, and hence as a work in progress. I'm constantly refining and improving it, but I haven't yet tested every single functionality. Hence, no guarantees regarding the actual version, but every guarantee about my willingness to constantly improve it.

Have fun,  
Will. 

Features
--------
+ Sync an entire Dropbox folder to a local folder. Both chosen by you.
+ Automatically performs folders creation, downloads, uploads, local and remote moves and deletions.
+ Has delegate! Therefore you'll have progress, loadMetadata start, sync start, sync change and sync completion. I've also temporarily overwritten Dropbox SDK's delegates for the basic operations like download, upload and delete.
+ Uses NSUserDefaults to cache loadMetadata and to perform checks to avoid conflicts and to create a delta.
+ Automatically creates a delta with cached data. No need to use Dropbox SDK's loadDelta.
+ Filter: escape root files both in Dropbox and local.
+ Filter: sync first level folders only, like iCloud's file system.
+ Filter: sync files only. At the moments syncs only files in the root.
+ Filter: sync images only. It will go through all folders and sync images only. The extensions supported are jpeg, jpg, png but you can obviously add whatever you want to.
+ Filter: sync thumbnails only. Whether you sync files only or images only, or everything, you can sync thumbnails only to speed up your app's responsiveness. You can later download the original image because we cache its original path, so it's perfect for apps that need to load a lot of images.
+ Filter: only sync a set of file extensions.
+ Filter: thumbnailSize corresponds to Dropbox SDK's thumbnailSize value. So be sure the value is one of those accepted by Dropbox.
+ (Not so) Magically creates an object of a file stored locally to match the one stored on Dropbox. The class is DeebeeItem. You can use it to load data into your app without touching NSFileManager for even a second.
+ Provides a method to unlink Dropbox and consequently erase every local item.
+ ARC.  

To Do
--------
+ Implement a better way to handle sync errors. At the moment, if any error, sync never ends.
+ Better progress. At the moment you'll get a partial progress since Deebee returns the progress for folder that was loaded through loadMetadata.
+ IF lastAbstractLocalItems has items, AND currentAbstractLocalItems has no item?
+ IF lastAbstractLocalItems has items, AND currentAbstractLocalItems has no item OR Dropbox has no item?

The End
--------
Special thanks to Chris Hulbert, who helped me to get started with [his CHBgDropboxSync](https://github.com/chrishulbert/CHBgDropboxSync).

Changelog
0.1.1
- Fixed a bug with _escapeRootFiles: NO file would have been downloaded if YES.


Deebee.h
//
//  Deebee.h
//  v 0.3 ß
//
//  Created by Will Flagello. MIT License.
//  ----------------------------------------------------------------------------------------------------
//  Special thanks to Chris Hulbert and his CHBgDropboxSync.
//
//  github.com/flvgello/deebee
//
//  For Dropbox iOS SDK 1.3
//

#import <Foundation/Foundation.h>
#import <DropboxSDK/DropboxSDK.h>
#import "DeebeeItem.h"

@class DeebeeItem;
@protocol DeebeeDelegate;

@interface Deebee : NSObject <DBRestClientDelegate>

@property (assign, nonatomic) id <DeebeeDelegate> delegate;

@property (readwrite, strong, nonatomic) NSString           *localRootPath;
@property (readwrite, strong, nonatomic) NSString           *remoteRootPath;
@property (readwrite, assign, nonatomic) BOOL               escapeRootFiles;             // Escape files in your app's root folder. (Dropbox & Local)
@property (readwrite, assign, nonatomic) BOOL               syncFirstLevelOnly;          // Sync Root + First Level Folders Only.
@property (readwrite, assign, nonatomic) BOOL               syncFilesOnly;               // Sync All Files, No Folders
@property (readwrite, assign, nonatomic) BOOL               syncImagesOnly;              // Sync All Images, All Folders (First-Level Only)
@property (readwrite, assign, nonatomic) BOOL               syncThumbnailsOnly;          // Sync All Thumbnails, No Images
@property (readwrite, assign, nonatomic) NSArray            *syncableFileExtensions;     // Syncs Only the files with the extensions in this array.
@property (readwrite, strong, nonatomic) NSString           *thumbnailsSize;             // WARNING: MUST follow Dropbox's naming conventions for thumbnails.

- (void)initWithSync;
- (void)performSync;
- (void)unlinkDropboxAndEraseAllLocalData;

@end


@protocol DeebeeDelegate <NSObject>

- (void)didStartLoadingMetadata;
- (void)didLoadMetadataWithItemsToSync:(int)remainingItems;
- (void)didLoadMetadataFailWithError:(NSError *)error;
- (void)didStartSyncing;
- (void)didChangeSyncingWithItemsToSync:(int)remainingItems;
- (void)didEndSyncingWithResult:(NSString *)result andError:(NSError *)error;

@optional
- (void)didUploadItem:(DeebeeItem *)item;
- (void)didDownloadItem:(DeebeeItem *)item;
- (void)itemDownloadFailedWithError:(NSError *)error;
- (void)itemUploadFailedWithError:(NSError *)error;
- (void)remoteItemDeleteFailedWithError:(NSError *)error;
- (void)localItemDeleteFailedWithError:(NSError *)error;

@end


DeebeeItem.h
//
//  DeebeeItem.h
//  v 0.3 ß
//
//  Created by Will Flagello. MIT License.
//  ----------------------------------------------------------------------------------------------------
//  Special thanks to Chris Hulbert and his CHBgDropboxSync.
//
//  github.com/flvgello/deebee
//
//  For Dropbox iOS SDK 1.3
//

#import <Foundation/Foundation.h>
#import <DropboxSDK/DropboxSDK.h>
#import "Deebee.h"

@interface DeebeeItem : NSObject <NSCoding>
{
    NSMutableString *name;
    NSDate *lastModified;
    NSMutableString *path;
    NSMutableString *localPath;
    NSMutableString *rev;
    BOOL isDirectory;

    // Folder
    NSMutableArray *contents;
    
    // File
    BOOL hasThumbnail;
    NSMutableString *extension;
}

@property (strong, nonatomic) NSString *ID;
// Since there's no way to create a unique id, we use the path.
// It lets us verify changes correctly even it's more like a workaround.

@property (readwrite, strong, nonatomic) NSMutableString *name;
@property (readwrite, strong, nonatomic) NSDate *lastModified;
@property (readwrite, strong, nonatomic) NSMutableString *path;             // Remote Path. It's great to idendify the important part of the local path.
@property (readwrite, strong, nonatomic) NSMutableString *localPath;        // Complete Local Path.
@property (readwrite, strong, nonatomic) NSMutableString *rev;
@property (readwrite, assign, nonatomic) BOOL isDirectory;

// Folder
@property (readwrite, strong, nonatomic) NSMutableArray *contents;

// File
@property (readwrite, assign, nonatomic) BOOL hasThumbnail;
@property (readwrite, strong, nonatomic) NSMutableString *extension;


- (id)initWithDictionary:(NSDictionary *)dictionary;
+ (DeebeeItem *)initFromMetadata:(DBMetadata *)metadata withLocalPath:(NSMutableString *)itemLocalPath;
+ (DeebeeItem *)initFromLocalItemAtPath:(NSString *)path withLastModifiedDate:(NSDate *)lastModified andLocalRootPath:(NSString *)localRootPath;

@end


DeebeeItem.m
//
//  DeebeeItem.m
//  v 0.3 ß
//
//  Created by Will Flagello. MIT License.
//  ----------------------------------------------------------------------------------------------------
//  Special thanks to Chris Hulbert and his CHBgDropboxSync.
//
//  github.com/flvgello/deebee
//
//  For Dropbox iOS SDK 1.3
//

#import "DeebeeItem.h"

@implementation DeebeeItem

+ (NSDateFormatter *)dateFormatter
{
    NSMutableDictionary *dictionary = [[NSThread currentThread] threadDictionary];
    static NSString *dateFormatterKey = @"DeebeeFolderDateFormatter";
    
    NSDateFormatter *dateFormatter = [dictionary objectForKey:dateFormatterKey];
    if (dateFormatter == nil){
        dateFormatter = [NSDateFormatter new];
        dateFormatter.locale = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US_POSIX"];
        dateFormatter.dateFormat = @"EEE, dd MMM yyyy HH:mm:ss Z";
        [dictionary setObject:dateFormatter forKey:dateFormatterKey];
    }
    
    return dateFormatter;
}
 
- (id)initWithDictionary:(NSDictionary *)dictionary
{
    if ((self = [super init])){
        if ([dictionary objectForKey:@"modified"])
            lastModified = [[DeebeeItem dateFormatter] dateFromString:[dictionary objectForKey:@"modified"]];
        
        ID = [dictionary objectForKey:@"ID"];

        path = [dictionary objectForKey:@"path"];
        localPath = [dictionary objectForKey:@"localPath"];
        rev = [dictionary objectForKey:@"rev"];
        
        extension = [dictionary objectForKey:@"extension"];
        hasThumbnail = [[dictionary objectForKey:@"hasThumbnail"] boolValue];
        isDirectory = [[dictionary objectForKey:@"isDirectory"] boolValue];

        if ([dictionary objectForKey:@"contents"]){
            NSArray *contentsOfFolder = [dictionary objectForKey:@"contents"];
            NSMutableArray *mutableFiles = [[NSMutableArray alloc] initWithCapacity:[contentsOfFolder count]];
            for (NSDictionary *item in contentsOfFolder){
                DeebeeItem *content = [[DeebeeItem alloc] initWithDictionary:item];
                [mutableFiles addObject:content];
            }
            contentsOfFolder = mutableFiles;
        }
    }
    return self;
}

- (BOOL)isEqual:(id)object
{
    if (object == self) return YES;
    if (![object isKindOfClass:[DeebeeItem class]]) return NO;
    
    DeebeeItem *other = (DeebeeItem *)object;
    return [self.ID isEqualToString:other.ID];
}

- (NSMutableString *)name
{
    if (name == nil)
        name = [NSMutableString stringWithString:[path lastPathComponent]];
    
    return name;
}

- (NSMutableString *)extension
{
    if (extension == nil && contents == nil)
        extension = [NSMutableString stringWithString:[path pathExtension]];
    
    return extension;
}

+ (DeebeeItem *)initFromMetadata:(DBMetadata *)metadata withLocalPath:(NSMutableString *)itemLocalPath
{
    DeebeeItem *item = [[DeebeeItem alloc] init];
    item.ID = metadata.path;
    item.path = [NSMutableString stringWithString:metadata.path];
    item.rev = [NSMutableString stringWithString:metadata.rev];
    item.name = [NSMutableString stringWithString:metadata.filename]; // Check if it has extension or not
    item.localPath = itemLocalPath;
    item.lastModified = metadata.lastModifiedDate;
    item.hasThumbnail = metadata.thumbnailExists;
    item.isDirectory = metadata.isDirectory;

    if (metadata.isDirectory){
        item.contents = [NSMutableArray arrayWithArray:metadata.contents];
        item.extension = nil;
    } else {
        item.contents = nil;
        item.extension = [NSMutableString stringWithString:[[metadata.path pathExtension] lowercaseString]];
    }
    
    return item;
}

+ (DeebeeItem *)initFromLocalItemAtPath:(NSString *)path withLastModifiedDate:(NSDate *)lastModified andLocalRootPath:(NSString *)localRootPath
{
    DeebeeItem *item = [[DeebeeItem alloc] init];
    item.ID = path;
    item.lastModified = lastModified;
    item.name = [NSMutableString stringWithString:[path lastPathComponent]];
    item.path = [NSMutableString stringWithString:[path stringByReplacingOccurrencesOfString:localRootPath withString:@""]];
    item.localPath = [NSMutableString stringWithString:path];
    item.rev = nil;
    
    NSDictionary *attribs = [[NSFileManager defaultManager] attributesOfItemAtPath:path error:nil];
    BOOL isPathDirectory = [attribs.fileType isEqual:NSFileTypeDirectory];
    BOOL isPathFile = [attribs.fileType isEqual:NSFileTypeRegular];
    
    item.isDirectory = isPathDirectory;
    if (isPathDirectory) item.extension = nil;
    if (isPathFile) item.extension = [NSMutableString stringWithString:[[path pathExtension] lowercaseString]];
    
    if (item.extension){
        item.contents = nil;
        NSArray *validImageExtensions = [NSArray arrayWithObjects:@"jpg", @"jpeg", @"png", nil];
        if ([validImageExtensions indexOfObject:item.extension] != NSNotFound){
            item.hasThumbnail = YES;
        } else {
            item.hasThumbnail = NO;
        }
    }
    
    return item;
}

@synthesize name;
@synthesize path;
@synthesize localPath;
@synthesize ID;
@synthesize contents;
@synthesize rev;
@synthesize isDirectory;
@synthesize extension;
@synthesize lastModified;
@synthesize hasThumbnail;

#pragma mark -------------------------------------------------------------------------------
#pragma mark NSCoding Methods
#pragma mark -------------------------------------------------------------------------------

- (id)initWithCoder:(NSCoder *)coder
{
    if ((self = [super init])){
        lastModified = [coder decodeObjectForKey:@"lastModified"];
        path = [coder decodeObjectForKey:@"path"];
        localPath = [coder decodeObjectForKey:@"localPath"];
        contents = [coder decodeObjectForKey:@"contents"];
        ID = [coder decodeObjectForKey:@"ID"];
        extension = [coder decodeObjectForKey:@"extension"];
        rev = [coder decodeObjectForKey:@"rev"];
        hasThumbnail = [coder decodeBoolForKey:@"hasThumbnail"];
        isDirectory = [coder decodeBoolForKey:@"isDirectory"];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
    [coder encodeObject:lastModified forKey:@"lastModified"];
    [coder encodeObject:path forKey:@"path"];
    [coder encodeObject:localPath forKey:@"localPath"];
    [coder encodeObject:contents forKey:@"contents"];
    [coder encodeObject:ID forKey:@"ID"];
    [coder encodeObject:extension forKey:@"extension"];
    [coder encodeObject:rev forKey:@"rev"];
    [coder encodeBool:hasThumbnail forKey:@"hasThumbnail"];
    [coder encodeBool:isDirectory forKey:@"isDirectory"];
}

@end
